### ISSUE TITLE
ETHSTAT: space in rrd names for some hosts only

### ISSUE THREAD

**Issue by TomGudman** [collectd/collectd#1059]
Hi,

The following behaviour is observed with collectd 5.4.2 and 5.5.0 (client and/or server). For an unknown reason yet it seems to only affect our database servers.

As you will see below, the path to the RRD file contains spaces.

```
[2015-06-03 16:43:42] uc_update: Value too old: name = xxx-mysqlxx.sub.domain.com.au/ethstat-eth0/derive-  TSO pkts tx; value time = 1433313813.000; last cache update = 1433313813.000;
[2015-06-03 16:43:43] uc_update: Value too old: name = db1.test.foo.bar.domain.com.au/ethstat-eth0/derive-  bcast bytes rx; value time = 1433313822.996; last cache update = 1433313822.996;
[2015-06-03 16:43:43] uc_update: Value too old: name = db1.test.foo.bar.domain.com.au/ethstat-eth0/derive-  pkts rx OOB; value time = 1433313822.996; last cache update = 1433313822.996;
[2015-06-03 16:43:43] uc_update: Value too old: name = db1.test.foo.bar.domain.com.au/ethstat-eth0/derive-  pkts rx err; value time = 1433313822.996; last cache update = 1433313822.996;
[2015-06-03 16:43:43] uc_update: Value too old: name = xxx-mysqlxx.sub.domain.com.au/ethstat-eth0/derive-  ucast bytes tx; value time = 1433313823.000; last cache update = 1433313823.000;
[2015-06-03 16:43:43] uc_update: Value too old: name = xxx-mysqlxx.sub.domain.com.au/ethstat-eth0/derive-  pkts linearized; value time = 1433313823.000; last cache update = 1433313823.000;
[2015-06-03 16:43:43] uc_update: Value too old: name = xxx-mysqlxx.sub.domain.com.au/ethstat-eth0/derive-  LRO byte rx; value time = 1433313823.000; last cache update = 1433313823.000;
[2015-06-03 16:43:43] uc_update: Value too old: name = db2.test.foo.bar.domain.com.au/ethstat-eth0/derive-  giant hdr; value time = 1433313813.095; last cache update = 1433313813.095;
[2015-06-03 16:43:43] uc_update: Value too old: name = db2.foo.bar.domain.com.au/ethstat-eth0/derive-  pkts linearized; value time = 1433313813.392; last cache update = 1433313813.392;
[2015-06-03 16:43:43] uc_update: Value too old: name = db2.foo.bar.domain.com.au/ethstat-eth0/derive-  pkts tx err; value time = 1433313813.392; last cache update = 1433313813.392;
```
```
# ls /var/lib/collectd/rrd/xxx-mysqlxx.sub.domain.com.au/ethstat-eth0/
derive-  bcast bytes rx.rrd        derive-     giant hdr.rrd    derive-  mcast pkts rx.rrd    derive-  rx buf alloc fail.rrd  derive-  ucast bytes rx.rrd
derive-  bcast bytes tx.rrd        derive-  giant hdr.rrd       derive-  mcast pkts tx.rrd    derive-Rx Queue#.rrd            derive-  ucast bytes tx.rrd
derive-  bcast pkts rx.rrd         derive-  hdr cloned.rrd      derive-  pkts linearized.rrd  derive-     too many frags.rrd  derive-  ucast pkts rx.rrd
derive-  bcast pkts tx.rrd         derive-     hdr err.rrd      derive-  pkts rx err.rrd      derive-  TSO bytes tx.rrd       derive-  ucast pkts tx.rrd
derive-  drv dropped rx total.rrd  derive-  LRO byte rx.rrd     derive-  pkts rx OOB.rrd      derive-  TSO pkts tx.rrd
derive-  drv dropped tx total.rrd  derive-  LRO pkts rx.rrd     derive-  pkts tx discard.rrd  derive-     tso.rrd
derive-     err.rrd                derive-  mcast bytes rx.rrd  derive-  pkts tx err.rrd      derive-Tx Queue#.rrd
derive-     fcs.rrd                derive-  mcast bytes tx.rrd  derive-  ring full.rrd        derive-tx timeout count.rrd
```

These servers have a ethstat config as per below:
```
LoadPlugin ethstat
<Plugin ethstat>
  Interface "eth0"
  Map "rx_csum_offload_errors" "if_rx_errors" "checksum_offload"
  Map "multicast" "if_multicast"
  MappedOnly false
</Plugin>
```

**Comment by rubenk** [comment]
Hi,

Interesting, it looks like the type_instance contains leading spaces.

Not sure if you still have these servers, but can you tell which nics they have, and dump the output of ethtool -S eth0?

We do an ioctl, SIOCETHTOOL on eth0 with the ETHTOOL_GSTRINGS cmd, it looks like in your case the kernel returns the stat names with spaces prepended.

**Comment by TomGudman** [comment]
Hi rubenk,

Yes, I still have the problem. They are all VMware ESXi 5.x VMs using the VMNet3 driver.

```
# ethtool -S eth0
NIC statistics:
     Tx Queue#: 0
       TSO pkts tx: 403388
       TSO bytes tx: 1308532250
       ucast pkts tx: 52364227408
       ucast bytes tx: 9496835560596
       mcast pkts tx: 16439623
       mcast bytes tx: 1084218702
       bcast pkts tx: 0
       bcast bytes tx: 0
       pkts tx err: 0
       pkts tx discard: 0
       drv dropped tx total: 0
          too many frags: 0
          giant hdr: 0
          hdr err: 0
          tso: 0
       ring full: 7
       pkts linearized: 0
       hdr cloned: 0
       giant hdr: 0
     Tx Queue#: 1
       TSO pkts tx: 839069
       TSO bytes tx: 2534053664
       ucast pkts tx: 71252895093
       ucast bytes tx: 90790874210454
       mcast pkts tx: 0
       mcast bytes tx: 0
       bcast pkts tx: 0
       bcast bytes tx: 0
       pkts tx err: 0
       pkts tx discard: 0
       drv dropped tx total: 0
          too many frags: 0
          giant hdr: 0
          hdr err: 0
          tso: 0
       ring full: 840
       pkts linearized: 0
       hdr cloned: 0
       giant hdr: 0
     Rx Queue#: 0
       LRO pkts rx: 0
       LRO byte rx: 0
       ucast pkts rx: 61069323477
       ucast bytes rx: 49587302358827
       mcast pkts rx: 32743205
       mcast bytes rx: 2161052426
       bcast pkts rx: 690150
       bcast bytes rx: 38686752
       pkts rx OOB: 2284103
       pkts rx err: 0
       drv dropped rx total: 0
          err: 0
          fcs: 0
       rx buf alloc fail: 0
     Rx Queue#: 1
       LRO pkts rx: 0
       LRO byte rx: 0
       ucast pkts rx: 62902118049
       ucast bytes rx: 50083532580398
       mcast pkts rx: 97472274
       mcast bytes rx: 133203703545
       bcast pkts rx: 0
       bcast bytes rx: 0
       pkts rx OOB: 2711993
       pkts rx err: 0
       drv dropped rx total: 0
          err: 0
          fcs: 0
       rx buf alloc fail: 0
     tx timeout count: 0

```

**Comment by TomGudman** [comment]
That's how it looks like in the collectd server logs for the same client host than the previous comment

```
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  giant hdr.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  ring full.rrd, [1465171721:840], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  bcast bytes tx.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  ucast bytes tx.rrd, [1465171721:90786720170267], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  mcast pkts tx.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-Rx Queue#.rrd, [1465171721:1], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  LRO pkts rx.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  LRO byte rx.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  ucast bytes rx.rrd, [1465171721:50081295478190], 1) failed with status -1. 
[2016-06-06 10:20:25] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  hdr cloned.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  mcast pkts rx.rrd, [1465171721:97466492], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  bcast pkts rx.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  bcast bytes rx.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  mcast bytes rx.rrd, [1465171721:133195801993], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-     err.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  pkts rx err.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  pkts rx OOB.rrd, [1465171721:2711993], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  rx buf alloc fail.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  ucast pkts rx.rrd, [1465171721:62899126072], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-  drv dropped rx total.rrd, [1465171721:0], 1) failed with status -1. 
[2016-06-06 10:20:36] rrdcached plugin: rrdc_update (/var/lib/collectd/rrd/xxx.sub.domain.tld/ethstat-eth0/derive-     fcs.rrd, [1465171721:0], 1) failed with status -1. 
```

**Comment by TomGudman** [comment]
Looks like you are right, the output of ethtool -S eth0 shows the exact amount of indent / spaces we can see in the file names. So I guess the patch is to trim any spaces to avoid this kind of issues. I am happy to try any patches. Cheers.

**Comment by rubenk** [comment]
Thanks for the update Tom.

**Comment by TomGudman** [comment]
Please have a look at #1756 

PS: I based my patch on the master branch instead of the latest 5.5 branch. Sorry.

**Comment by rubenk** [comment]
Merged as 21602de

**Comment by rsommer** [comment]
Hi, the mentioned fix does indeed correct the whitespace problem. The shown output of ethstat reveals another problem: the multiple queues available since VMXnet3 driver 1.0.16.0-k and later (introduced in Linux kernel 2.6.37) seem to be the source of the `uc_update: Value too old:`-errors, because the various metrics are reported for every queue, but the queue itself is not part of the metric name. This is generating a name conflict which leads to the often mentioned `uc_update: Value too old:` log entries.
